File: android/sdk/src/main/kotlin/com/pixeldive/sdk/ProtoWire.kt
package com.pixeldive.sdk

internal object ProtoWire {
    fun tag(
        field: Int,
        wire: Int,
    ): ByteArray = varint((field shl 3 or wire).toLong())

    fun varint(value: Long): ByteArray {
        val out = ArrayList<Byte>(10)
        var current = value
        while (true) {
            if (current and 0x7FL.inv() == 0L) {
                out.add(current.toByte())
                return out.toByteArray()
            }
            out.add(((current and 0x7F) or 0x80).toByte())
            current = current ushr 7
        }
    }

    fun stringField(
        field: Int,
        value: String,
    ): ByteArray = bytesField(field, value.toByteArray(Charsets.UTF_8))

    fun bytesField(
        field: Int,
        value: ByteArray,
    ): ByteArray = tag(field, 2) + varint(value.size.toLong()) + value

    fun varintField(
        field: Int,
        value: Long,
    ): ByteArray = tag(field, 0) + varint(value)

    fun boolField(
        field: Int,
        value: Boolean,
    ): ByteArray = if (value) varintField(field, 1) else ByteArray(0)
}

internal class ProtoReader(
    private val bytes: ByteArray,
) {
    var offset = 0
        private set

    fun next(): Triple<Int, Int, ByteArray>? {
        if (offset >= bytes.size) {
            return null
        }
        val key = readVarint()
        val field = (key ushr 3).toInt()
        val wire = (key and 0x7).toInt()
        return Triple(field, wire, readWire(wire))
    }

    fun readVarint(): Long {
        var result = 0L
        var shift = 0
        while (true) {
            if (offset >= bytes.size) {
                throw PixeldiveException.Decoding("truncated protobuf varint")
            }
            val byte = bytes[offset].toInt() and 0xFF
            offset += 1
            result = result or ((byte and 0x7F).toLong() shl shift)
            if (byte and 0x80 == 0) {
                return result
            }
            shift += 7
            if (shift > 63) {
                throw PixeldiveException.Decoding("protobuf varint overflow")
            }
        }
    }

    private fun readWire(wire: Int): ByteArray =
        when (wire) {
            0 -> {
                val start = offset
                readVarint()
                bytes.copyOfRange(start, offset)
            }
            1 -> read(8)
            2 -> read(readVarint().toInt())
            5 -> read(4)
            else -> throw PixeldiveException.Decoding("unsupported protobuf wire type $wire")
        }

    private fun read(count: Int): ByteArray {
        val end = offset + count
        if (count < 0 || end > bytes.size) {
            throw PixeldiveException.Decoding("truncated protobuf field")
        }
        val slice = bytes.copyOfRange(offset, end)
        offset = end
        return slice
    }
}

internal fun decodeVarint(data: ByteArray): Long = ProtoReader(data).readVarint()

File: ios/Sources/PixeldiveSDK/ProtoWire.swift
import Foundation

enum ProtoWire {
    static func tag(_ field: Int, wire: UInt8) -> Data {
        varint(UInt64(field << 3 | Int(wire)))
    }

    static func varint(_ value: UInt64) -> Data {
        var data = Data()
        var current = value
        repeat {
            var byte = UInt8(current & 0x7F)
            current >>= 7
            if current != 0 {
                byte |= 0x80
            }
            data.append(byte)
        } while current != 0
        return data
    }

    static func stringField(_ field: Int, _ value: String) -> Data {
        bytesField(field, Data(value.utf8))
    }

    static func bytesField(_ field: Int, _ value: Data) -> Data {
        var data = tag(field, wire: 2)
        data.append(varint(UInt64(value.count)))
        data.append(value)
        return data
    }

    static func varintField(_ field: Int, _ value: UInt64) -> Data {
        var data = tag(field, wire: 0)
        data.append(varint(value))
        return data
    }

    static func boolField(_ field: Int, _ value: Bool) -> Data {
        value ? varintField(field, 1) : Data()
    }
}

struct ProtoField {
    var number: Int
    var wire: UInt8
    var value: Data
}

struct ProtoReader {
    let bytes: Data
    var offset = 0

    mutating func next() throws -> ProtoField? {
        if offset >= bytes.count {
            return nil
        }
        let key = try readVarint()
        let field = Int(key >> 3)
        let wire = UInt8(key & 0x7)
        switch wire {
        case 0:
            let start = offset
            _ = try readVarint()
            return ProtoField(number: field, wire: wire, value: bytes.subdata(in: start..<offset))
        case 1:
            return ProtoField(number: field, wire: wire, value: try read(8))
        case 2:
            let length = Int(try readVarint())
            return ProtoField(number: field, wire: wire, value: try read(length))
        case 5:
            return ProtoField(number: field, wire: wire, value: try read(4))
        default:
            throw PixeldiveError.decoding("unsupported protobuf wire type \(wire)")
        }
    }

    mutating func readVarint() throws -> UInt64 {
        var result: UInt64 = 0
        var shift = 0
        while true {
            if offset >= bytes.count {
                throw PixeldiveError.decoding("truncated protobuf varint")
            }
            let byte = bytes[offset]
            offset += 1
            result |= UInt64(byte & 0x7F) << shift
            if byte & 0x80 == 0 {
                return result
            }
            shift += 7
            if shift > 63 {
                throw PixeldiveError.decoding("protobuf varint overflow")
            }
        }
    }

    mutating func read(_ count: Int) throws -> Data {
        let end = offset + count
        if count < 0 || end > bytes.count {
            throw PixeldiveError.decoding("truncated protobuf field")
        }
        let slice = bytes.subdata(in: offset..<end)
        offset = end
        return slice
    }
}

func decodeVarint(_ data: Data) throws -> UInt64 {
    var reader = ProtoReader(bytes: data)
    return try reader.readVarint()
}
